### DISCUSSION TITLE
How to handle line returns in grammar?

### DISCUSSION THREAD

**Discussion by ShirlyWang666** [eclipse-langium/langium#971] [Q&A] [answered]
Is it possible to parse line return to a specific AstNode? Let's say, when user hits enter and creates a new line, I want to do something specific, like generate another object according to the AstNode. 

In current version, If I use '\n' as the keyword, there is an error message popped up 'Keywords cannot only consist of whitespace characters.'. So, is there a way to handle this case?

**Comment by msujew** [comment, marked as answer]
You can in theory use `\n` in a terminal and use it in the AST:

```
entry Test: return=RETURN;

terminal RETURN: /\r?\n/;
```

However, you must adjust the `hidden terminal WS` in order for this to work:

```
hidden terminal WS: /[ \t\r]+/;
```

At this point, you will need to track every `\n` character manually in your grammar, which is very annoying, so I wouldn't recommend it. Why is it that you need a special AST node? This would make the grammar insanely complicated, since most languages allow newlines basically anywhere.

**Reply by ShirlyWang666** [reply]
@msujew Thanks so much for your kindly suggestion. I totally agree with you, it's very annoying to handle '\n'. I will think it over.

**Reply by msujew** [reply]
By the way, you can always look into the CST (concrete syntax tree, available under the `$cstNode` property of each `AstNode`) to see whether there is a line break between two CST nodes. Incorporating whitespace information in the AST is usually not a good idea and will lead to issues somewhere down the line.

**Reply by ShirlyWang666** [reply]
@msujew Thanks a million for the tips.  I was wondering whether I need to start a new discussion. Actually, I am trying to leverage line return to generate AstNode for a line, but I found out It does not work for the first line. Any advice is appreciated.
